### PR TITLE
Added support for IRB4600 20/2.50 and 40/2.55

### PR DESCRIPTION
**Pull request by rtonnaer**
this adds support for the IRB4600 20/2.50 and 40/2.55 variants. these variants only have minor differences from the IRB4600 that was already supported.

### CODE REVIEW

**Comment by gavanderhoorn** [comment]
Thanks for the PR.

I haven't checked yet, but if these variants share any link geometry (ie: meshes) then it would be nice if we could avoid adding duplicate `.stl`s for those that can be shared.

Do you see any opportunity to do that?

**Comment by rtonnaer** [comment]
Yes that would be possible. I see multiple ways of doing this. How should I sent up the /meshes folder?

For instance add common/visual and common/collision folders and put common stls in these folders only. This would mean that the irb4600_xx_xxx/visual and irb4600_xx_xxx/collision folders only contain stls for unique links.

Another option is to use one irb4600_xx_xx variant as base and refer to those stls for the other variants.

**Comment by gavanderhoorn** [comment]
A `meshes/common` folder is a possibility. This was done for the IRB 2400 ([here](https://github.com/ros-industrial/abb/tree/ca72a6276a422bb75f361523ab50bd5a020dd0b3/abb_irb2400_support/meshes)).

> Another option is to use one irb4600_xx_xx variant as base and refer to those stls for the other variants.

If there are really only minor differences between the variants, then using one as the base would be perfectly fine.

I haven't checked this specific model, so if you could provide an estimate / description of the similarities?

**Comment by RyanDelGizzi** [comment]
I am beginning a project requiring the 4600_40_255.  @rtonnaer is there anything I can do to help get the PR merged?

**Comment by rtonnaer** [comment]
Well, I think only base_link and link1 is exactly similar over all models in the 4600 range. It might be that two models share the same link2. But I think moving these (partially) similar geometries to the `meshes/common` folder makes the package less structured. This would also cause more difficult link names. @gavanderhoorn do you agree?

@RyanDelGizzi can you verify that `irb4600_40_255_macro.xacro` is correct/works for you?

**Comment by RyanDelGizzi** [comment]
> @RyanDelGizzi can you verify that `irb4600_40_255_macro.xacro` is correct/works for you?

Yes, it works for me and is correct as far as I can tell.  

I have also made changes on my end to support gazebo.  These changes are pretty simple, and are directly from the moveit setup assistant.  They are similar to the [urdfs](https://github.com/ros-industrial/abb_experimental/tree/kinetic-devel/abb_irb120_support/urdf) in the abb_irb120_support package.  @rtonnaer if you want to make these changes, go ahead.  If you'd like for me to submit a pull request to you, I can do that as well.  If you'd rather just get the PR through, then I'll submit one to the main repository in the future.

**Comment by RyanDelGizzi** [comment]
@rtonnaer the stl files for the 4600 visual produce a warning in rviz that they are malformed.

```sh
ros.rviz: The STL file 'package://abb_irb4600_support/meshes/irb4600_40_255/visual/link_1.stl' is malformed. It starts with the word 'solid', indicating that it's an ASCII STL file, but it does not contain the word 'endsolid' so it is either a malformed ASCII STL file or it is actually a binary STL file. Trying to interpret it as a binary STL file instead.
```
Importing the file to blender and then re-exporting the file removes the error.

**Comment by gavanderhoorn** [comment]
> I have also made changes on my end to support gazebo. These changes are pretty simple, and are directly from the moveit setup assistant. They are similar to the [urdfs](https://github.com/ros-industrial/abb_experimental/tree/kinetic-devel/abb_irb120_support/urdf) in the abb_irb120_support package. @rtonnaer if you want to make these changes, go ahead. If you'd like for me to submit a pull request to you, I can do that as well. If you'd rather just get the PR through, then I'll submit one to the main repository in the future.

Note that the MSA exports a `.urdf`, while the files in robot support packages are `.xacro`s.

The changes required for Gazebo *can* be applied to `.xacro`s as well, but that would have to be done manually.

The `.urdf`s in `abb_irb120_support` should actually be deleted: they are not needed and increase the maintenance burden on the maintainers.

**Comment by RyanDelGizzi** [comment]
> Note that the MSA exports a `.urdf`, while the files in robot support packages are `.xacro`s.
> 
> The changes required for Gazebo _can_ be applied to `.xacro`s as well, but that would have to be done manually.

Yes, sorry for the confusion. I meant I manually applied the changes to the `.xacro` file.  The MSA simply tells you what changes to make, I do not believe it actually changes the file. 
@gavanderhoorn do you think we should apply these changes in this PR or hold off until these changes are merged?

**Comment by gavanderhoorn** [comment]
> @gavanderhoorn do you think we should apply these changes in this PR or hold off until these changes are merged?

Let's add them in a follow-up PR.

**Comment by rtonnaer** [comment]
> @rtonnaer the stl files for the 4600 visual produce a warning in rviz that they are malformed.
> 
> ```shell
> ros.rviz: The STL file 'package://abb_irb4600_support/meshes/irb4600_40_255/visual/link_1.stl' is malformed. It starts with the word 'solid', indicating that it's an ASCII STL file, but it does not contain the word 'endsolid' so it is either a malformed ASCII STL file or it is actually a binary STL file. Trying to interpret it as a binary STL file instead.
> ```
> 
> Importing the file to blender and then re-exporting the file removes the error.

Thanks for pointing this out. Due to (I think) the logging level, I did not see these warnings before.

Apparently it is a common issue when exporting from Solidworks ([question on ROS answers](https://answers.ros.org/question/11305/how-to-generate-stl-files/)). I have solved this using [this comment](https://github.com/ros-industrial/motoman/issues/123) from Gijs on another Issue.

**Comment by rtonnaer** [comment]
okay I didn't mean to close this Pull Request. So I opened it again.

**Comment by RyanDelGizzi** [comment]
Thanks @rtonnaer, the changes you made work for me.

File: abb_irb4600_support/urdf/irb4600_20_250_macro.xacro
Comment (gavanderhoorn) [inline abb_irb4600_support/urdf/irb4600_20_250_macro.xacro]: Please use `${radians(..)}` to convert from degrees to radians. You can use the specsheet to find the joint limits in degrees (which you probably also used to convert things manually here).

File: abb_irb4600_support/urdf/irb4600_20_250_macro.xacro
Comment (gavanderhoorn) [inline abb_irb4600_support/urdf/irb4600_20_250_macro.xacro:140]: This transform has non-zero elements for two axes (X and Z). We prefer to limit transforms for axes in robots to a translation in a single dimension only.

Compare `joint_4` in [irb4600_60_205_macro.xacro](https://github.com/ros-industrial/abb_experimental/blob/d6900d8b56913a10558db6ca1e0e6a68d7896690/abb_irb4600_support/urdf/irb4600_60_205_macro.xacro#L123-L129).

Note: this will require a translation of the mesh as well.

File: abb_irb4600_support/urdf/irb4600_20_250_macro.xacro
Comment (gavanderhoorn) [inline abb_irb4600_support/urdf/irb4600_20_250_macro.xacro]:
```suggestion
      <origin xyz="0 0 0" rpy="0 ${pi/2.0} 0"/>
```

File: abb_irb4600_support/urdf/irb4600_40_255_macro.xacro
Comment (gavanderhoorn) [inline abb_irb4600_support/urdf/irb4600_40_255_macro.xacro]: Same comment re: `radians` here.

File: abb_irb4600_support/urdf/irb4600_40_255_macro.xacro
Comment (gavanderhoorn) [inline abb_irb4600_support/urdf/irb4600_40_255_macro.xacro:140]: Same comment re: translation in 2 dimensions.

File: abb_irb4600_support/urdf/irb4600_40_255_macro.xacro
Comment (gavanderhoorn) [inline abb_irb4600_support/urdf/irb4600_40_255_macro.xacro]:
```suggestion
      <origin xyz="0 0 0" rpy="0 ${pi/2.0} 0"/>
```

**Review by gavanderhoorn — CHANGES_REQUESTED** [review]
Some minor comments, otherwise looks good.

Could you also please update `package.xml` to mention the added variants and `CMakeLists.txt` to execute the added `roslaunch` tests?

**Comment by RyanDelGizzi** [comment]
Hi @rtonnaer,  will you have a chance to make these changes in the near future?  If not, would you like me to make the changes and create pull request to your repo?

**Comment by gavanderhoorn** [comment]
@rtonnaer: :bellhop_bell: ?

**Comment by RyanDelGizzi** [comment]
@gavanderhoorn what is the best way for me to contribute the requested changes?

**Comment by gavanderhoorn** [comment]
> @gavanderhoorn what is the best way for me to contribute the requested changes?

thanks for the offer.

I'll first talk to @rtonnaer once more to see whether he can fix some of this (I believe he's identified an additional issue with the xacros in this PR).

As to contributing: a PR against @rtonnaer's fork (and branch) might be the most straightforward way. I'll make sure he reviews and merges it.

**Comment by RyanDelGizzi** [comment]
OK, I'll wait until next week to see if there is any progress made this week.  

I think I found an issue as well, which is why I wanted to get these first changes made. @rtonnaer there is no way to file an issue against your repo, so if you do make changes, the location of joint three should be 1.095 instead of 0.9 (at least on the 40-2.55.  If I submit a PR to you Ill include these changes.

Thanks!

**Comment by gavanderhoorn** [comment]
> I think I found an issue as well, which is why I wanted to get these first changes made. @rtonnaer there is no way to file an issue against your repo, so if you do make changes, the location of joint three should be 1.095 instead of 0.9 (at least on the 40-2.55. If I submit a PR to you Ill include these changes.

I think that's the issue @rtonnaer told me about.

You could just submit a PR in any case for that. It's a simple fix -- should be doable with the online GH editor.

**Comment by RyanDelGizzi** [comment]
It looks like @rtonnaer already made that change!  Thanks @gavanderhoorn for the coaching.

**Comment by gavanderhoorn** [comment]
:bellhop_bell: @rtonnaer.

**Comment by rtonnaer** [comment]
i've now updated all axis limits to ${radians(DEG_VALUE)}.

I won't get around to update the 2 translations issue today or tomorrow, as this involves moving the meshes themselves in their own coordinate system.

Will put it on my todo list for friday.

**Comment by gavanderhoorn** [comment]
@rtonnaer: thanks for the updates.

Do you know where f8e2bf903c0875b0799521ec383d15b76344c249 came from? That seems like a commit that should not be part of this PR.

**Comment by rtonnaer** [comment]
@RyanDelGizzi with the help of @gavanderhoorn I finished adding the ABB variants. For cleanliness we decided to make a new pull request (#112). Can you verify that this works for you?

**Pull request closed without merging**
